### PR TITLE
framework/microunit: add some basic structs

### PR DESCRIPTION
**Pull request by huachaohuang**
Wow, we have a runnable binary with this PR :)

```
USAGE:
    engula node start <ADDR>
```

### CODE REVIEW

File: src/engula-framework/microunit/src/node_server.rs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+/// An HTTP server that serves a node.
+pub struct NodeServer {}
+
+impl NodeServer {
+    pub fn bind(_addr: &SocketAddr) -> NodeServer {
+        NodeServer {}
+    }
+
+    pub fn serve(&self, _node: Node) {}
```
Comment (tisonkun) [inline src/engula-framework/microunit/src/node_server.rs:27]: Aha. Dummy implementation.

File: src/engula-framework/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+[package]
+name = "engula-framework"
+version = "0.1.0"
+edition = "2021"
+
+[dependencies]
+microunit = { path = "microunit" }
```
Comment (PsiACE) [inline src/engula-framework/Cargo.toml:7]: It seems that two modules should be avoided to depend on each other

**Review by tisonkun — CHANGES_REQUESTED** [review]
Can we try to adopt a two level ident package structure and let microunit stay at `src/microunit`?

From https://github.com/engula/engula/pull/62#discussion_r742488203 I found it's hard to distinguish engula (the storage engine) and microunit (which should be unaware of the storage engine logic) if microunit stays at `src/engula-framework/microunit` which leads to an imagine microunit being part of engula.

Also `engula-` prefix seems redundant. You can take [materialize](https://github.com/MaterializeInc/materialize/tree/main/src) as an example.

**Comment by huachaohuang** [comment]
> Can we try to adopt a two level ident package structure and let microunit stay at `src/microunit`?
> 
> From [#62 (comment)](https://github.com/engula/engula/pull/62#discussion_r742488203) I found it's hard to distinguish engula (the storage engine) and microunit (which should be unaware of the storage engine logic) if microunit stays at `src/engula-framework/microunit` which leads to an imagine microunit being part of engula.

So from the perspective of Engula, microunit is part of it, no matter whether microunit is in another crate or repository. However, inside the microunit, it is a standalone crate that can be published alone or moved to another repository.

> 
> Also `engula-` prefix seems redundant. You can take [materialize](https://github.com/MaterializeInc/materialize/tree/main/src) as an example.

A big difference between Engula and other projects like Materialize is that Engula is more a library than a binary. Materialize doesn't need to care about the name of internal crates because those crates are not going to be published. But for Engula, we may want to publish the top-level crates (`engula-xxx` crates) and keep second-level crates internal and unpublished.

**Comment by huachaohuang** [comment]
> A big difference between Engula and other projects like Materialize is that Engula is more a library than a binary. Materialize doesn't need to care about the name of internal crates because those crates are not going to be published. But for Engula, we may want to publish the top-level crates (`engula-xxx` crates) and keep second-level crates internal and unpublished.

Another alternative is to keep only one top-level crate `engula` and make all other crates internal. This alternative is simpler although users have to compile all features of Engula together. Yet another trick we can do is to use conditional compilation with [feature flags](https://doc.rust-lang.org/cargo/reference/features.html).

**Comment by tisonkun** [comment]
@huachaohuang thanks for your explanation. Let me think once more about publishing.

**Comment by huachaohuang** [comment]
Let's move the project layout discussion to https://github.com/engula/engula/issues/59. I think the alternative is worth a try if you think it's easier to understand.

**Review by tisonkun — APPROVED** [review]
LGTM. Postpone layout discussion to #59 .

**Pull request merged**